pkg/dataobj/consumer/partition_processor.go

```diff
@@ -177,7 +193,19 @@ func (p *partitionProcessor) processRecord(record *kgo.Record) {
 			backoff.Wait()
 		}
 
-		if err := p.metastoreManager.UpdateMetastore(p.ctx, flushResult); err != nil {
+		backoff.Reset()
+		var objectPath string
+		for backoff.Ongoing() {
+			objectPath, err = p.uploader.Upload(p.ctx, p.flushBuffer)
+			if err == nil {
+				break
+			}
+			level.Error(p.logger).Log("msg", "failed to upload object", "err", err)
+			p.metrics.incUploadFailures()
```

Comment:
There are a few options when it comes to handling the metrics for this package which currently have split responsibilities.
* Partition_processor is responsible for reporting error counters during flush or uploading
* Timing metrics are reported by the uploader or builder itself
* Metastore manager is handling all it's own error count & timing metrics plus retries because it does several pieces of work due to potentially multiple metastores being updated.

I was tempted to remove the Uploader.metrics object in order to reduce the need for a new metrics object + register dance. The metrics would be handled by the PartitionProcessor but that sets up a third approach so I copied what we do for the builder to try to minimise confusion.

Reply:
I agree that the consumer metrics are encroaching on responsibilities that should be handled by the builder, uploader, or metastore. The only metrics that the partition processor _need_ to expose itself is anything Kafka-specific (lag, current offset).